### PR TITLE
case 20041: Package all the graphics related objects and functions in Application in the Graphics Engine

### PR DESCRIPTION
**Pull request by samcake**
Ticket: 20041
https://highfidelity.fogbugz.com/f/cases/20041/Refactor-the-render-and-present-components-of-Application-into-the-Graphics-Engine

This PR achieve 2 things:
1/ it cleans and fixes bugs in the task::Config object, 
    - The retrieval of a Job from its name or path in the graph of tasks is now correctly working. This is fixing bugs in the ui displaying the structure of an engine. 
    - Removed the "always Enabled" boolean attribute which is not useful and creates confusion
    - THe Jet script and qml library Now is able to represent and interact correctly with a top level engine such as Workload or Render.
     These changes have no consequences on the behavior of the engines, simply addressing issues related mostly to script and qml tools used by developers.
       https://highfidelity.fogbugz.com/f/cases/19140/Accessing-the-task-Job-Config-from-Scripting-interface-is-not-reliable-enough-in-qml-and-sometimes-crashes-let-s-fix-that

2/ It introduces a new class in  Interface,  the Graphics Engine which gathers in one place all the graphics (gpu, render and present) related functions and objects that were in Application.h.cpp
THis code is moving all the separate components into 2 classes and matching files:
   - GraphicsEngine .h/cpp is gathering the render engine, the render scene, and the functions spawning the render thread and call to actually spin the render engine.
    - RenderEventHandler .h/cpp is the same class as before now separated in its own file and interacting with Graphics Engine
    - WorldAxis which was is a piece of history in our code has been saved in its own file

## TEST PLAN
- run nitpick plan on windows and mac
- application performance should be identical
- gameplay rate should be identical to master

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15324/HighFidelity-Beta-PR14455-6b1c7bf.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15324/HighFidelity-Beta-Interface-PR14455-6b1c7bf.exe

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15328/HighFidelity-Beta-PR14455-48f32dc.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15328/HighFidelity-Beta-Interface-PR14455-48f32dc.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15328/HighFidelity-Beta-PR14455-48f32dc.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15328/HighFidelity-Beta-Interface-PR14455-48f32dc.exe

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15341/HighFidelity-Beta-PR14455-9ee35de.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15341/HighFidelity-Beta-PR14455-9ee35de.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15341/HighFidelity-Beta-Interface-PR14455-9ee35de.dmg

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15365/HighFidelity-Beta-PR14455-f9fb4b6.apk

**Comment by samcake** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15380/HighFidelity-Beta-PR14455-f9fb4b6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15380/HighFidelity-Beta-Interface-PR14455-f9fb4b6.exe

**Comment by samcake** [comment]
build this please

**Comment by samcake** [comment]
> @samcake is this ready for development cr?

Yes

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15444/HighFidelity-Beta-PR14455-f9fb4b6.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15444/HighFidelity-Beta-PR14455-f9fb4b6.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15444/HighFidelity-Beta-Interface-PR14455-f9fb4b6.exe

File: scripts/developer/utilities/lib/jet/qml/TaskListView.qml [hunk 1]
```diff
@@ -19,56 +19,43 @@ import "../jet.js" as Jet
...
+        Jet.task_traverseTree(rootConfig, tfunctor);
+*/
+      //  var currentParentStach = []
+    //    currentParentStach.push(jobsModel);
+        
 
+      /*  Jet.job_traverseTreeNodeRoot(jobsModel.engineJobItemModel[0], function(node, depth, index) {
+            print(node.name + depth + " - " + index)
+            return true
+        })*/
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/lib/jet/qml/TaskListView.qml]: Is this stuff cruft?  The comment indentation is a bit messed up.  If we wanna keep it maybe clean it up and add a comment about why we want it around?

File: scripts/developer/utilities/lib/jet/qml/TaskListView.qml [hunk 2]
```diff
@@ -77,30 +64,48 @@ Rectangle {
...
                         text: (objRecursiveColumn.children.length > 2 ?
                                 objRecursiveColumn.children[1].visible ?
-                                qsTr("-  ") : qsTr("+ ") : qsTr("   ")) + model.name + " ms=" + config.cpuRunTime.toFixed(2)
-                        checked: config.enabled
+                                qsTr("-  ") : qsTr("+ ") : qsTr("   ")) + model.name
+                              //  + " ms=" + config.cpuRunTime.toFixed(3)
+                                + " id=" + model.id
                     }
-                }
+                }
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/lib/jet/qml/TaskListView.qml]: New unused whitespace (not a blocker).

File: scripts/developer/utilities/lib/jet/qml/TaskTimeFrameView.qml [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+                x0 = getXFromTime(10)
+                ctx.moveTo(x0, 0); 
+                ctx.lineTo(x0, height); 
+
+                ctx.stroke();
+            }
+
+            function drawJob(ctx, depth, index, duration, timeOffset) {
+               //print(root.jobsArray[index].cpuT)
+              //  ctx.fillStyle = Qt.rgba(255, 255, 0, root.backgroundOpacity);
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/lib/jet/qml/TaskTimeFrameView.qml]: Crufty `ctx.fillStyle` setting?

File: scripts/developer/utilities/render/engineInspector.js [hunk 4]
```diff
@@ -1,13 +1,71 @@
...
+            width: 500, 
+            height: 100
         });
         window.setPosition(200, 50);
-        //window.closed.connect(function() { Script.stop(); });
+        window.closed.connect(function() { Script.stop(); });
     }
-    openEngineTaskView();
\ No newline at end of file
+    openEngineTaskView();*/
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/render/engineInspector.js]: Cruft or keeper?

File: scripts/developer/utilities/render/engineInspector.qml [hunk 5]
```diff
@@ -18,13 +18,13 @@ import "../lib/jet/qml" as Jet
...
-    property var mainViewTask: Render.getConfig("RenderMainView")
+    property var rootConfig: Render.getConfig("")
 
     Jet.TaskListView {
-        rootConfig: Render
-        anchors.fill: render        
-    }
+        rootConfig: root.rootConfig
+        anchors.fill: root        
+    }
```
Comment (AndrewMeadows) [inline scripts/developer/utilities/render/engineInspector.qml]: Unused whitespace (not a blocker).

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15519/HighFidelity-Beta-PR14455-54fa436.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15519/HighFidelity-Beta-PR14455-54fa436.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15519/HighFidelity-Beta-Interface-PR14455-54fa436.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15519/HighFidelity-Beta-PR14455-54fa436.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15519/HighFidelity-Beta-Interface-PR14455-54fa436.exe

**Comment by qaAdamSmith82** [comment]
Build this please

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15625/HighFidelity-Beta-PR14455-54fa436.apk

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15625/HighFidelity-Beta-PR14455-54fa436.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/15625/HighFidelity-Beta-Interface-PR14455-54fa436.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15625/HighFidelity-Beta-PR14455-54fa436.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15625/HighFidelity-Beta-Interface-PR14455-54fa436.exe

**Comment by qaAdamSmith82** [comment]
Fail @conklin94122 

This PR fails the Highlights tests in Nitpick.
https://github.com/highfidelity/hifi_tests/tree/master/tests/engine/render/effect/highlight
These are the results in Master (9693):
https://hifi-qa.s3.amazonaws.com/TestResults--2018-12-03_18-39-53(local)[2060-DESKTOP-PC]/TestResults.html
All tests pass

These are the results in PR14455
https://hifi-qa.s3.amazonaws.com/TestResults--2018-12-03_18-33-25(local)[2060-DESKTOP-PC]/TestResults.html

highlight

**Comment by hifi-gustavo** [comment]
Android build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/15637/HighFidelity-Beta-PR14455-d2ff4d4.apk

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15637/HighFidelity-Beta-PR14455-d2ff4d4.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/15637/HighFidelity-Beta-Interface-PR14455-d2ff4d4.exe

**Comment by samcake** [comment]
I think there is a bug in Nitpick that has been exposed with this pr.
THe secondary camera highlighting is  disabled but this is not working correctly in current hifi master.
this is why the highlighting tests show correctly in the reference snapshot of the test.

I think this PR is correct

and i have proposed a pr  that fixes Nitpick instead:
https://github.com/highfidelity/hifi_tests/pull/304
if we validate that the Highlight coverage test from that Hifi_tests PR304 is passing with this Hifi PR then that is proving the point.

 the 2 should be merged at the same time

**Comment by qaAdamSmith82** [comment]
Pass @conklin94122 

The Nitpick bug is fixed, and all rendering tests now pass.
QML tests pass.

**Pull request merged**